pkg/nsx/services/vpc/vpc.go

```diff
@@ -471,3 +523,236 @@ func (s *VPCService) Cleanup() error {
+			result = append(result, *ipblock.Cidr)
+		}
+	}
+	return
+}
+
+func (service *VPCService) CreateOrUpdateAVISecurityPolicy(vpc *model.Vpc, namespace string) error {
+	if !service.supportAviAllowRule {
+		return nil
+	}
+	path := strings.Split(*vpc.Path, "/")
+	orgId := path[2]
+	projectId := path[4]
+	ruleId := AllowAviRuleId
+	groupId := AllAviSEGroupId
+	spId := AviSecurityPolicyId
+
+	if !service.checkAVISecurityPolicyExist(orgId, projectId, *vpc.Id, spId) {
+		return errors.New("avi security policy not found")
+	}
+	allowrule, err := service.getAVIAllowRule(orgId, projectId, *vpc.Id, spId, ruleId)
+	var externalCIDR []string
+	if err == nil {
+		externalCIDR, err = service.getIpblockCidr(vpc.ExternalIpv4Blocks)
+		if err != nil {
+			return err
+		}
+	}
+
+	if allowrule != nil {
+		if !service.needUpdateRule(allowrule, externalCIDR) {
+			log.Info("avi rule not changed, skip update avi rule")
+			return nil
+		} else {
+			log.Info("avi rule changed", "previous", allowrule.DestinationGroups, "current", externalCIDR)
+		}
+	}
+
+	group, err := service.getorCreateAVIGroup(orgId, projectId, *vpc.Id, groupId)
+	if err != nil {
+		log.Error(err, "failed to get avi group", "group", groupId)
+		return err
+	}
+
+	newrule, err := service.buildAVIAllowRule(vpc, externalCIDR, *group.Path, ruleId, projectId)
+	log.Info("creating avi rule", "rule", newrule)
+	if err != nil {
+		log.Error(err, "build avi rule", "rule", newrule)
+		return err
+	}
+
+	err = service.NSXClient.VPCRuleClient.Patch(orgId, projectId, *vpc.Id, spId, *newrule.Id, *newrule)
+	if err != nil {
+		log.Error(err, "create avi rule", "rule", newrule)
+		return err
+	}
+	nsxrule, err := service.NSXClient.VPCRuleClient.Get(orgId, projectId, *vpc.Id, spId, *newrule.Id)
+	if err != nil {
+		log.Error(err, "get avi rule", "rule", nsxrule)
+		return err
+	}
+	service.RuleStore.Add(nsxrule)
+	log.Info("created avi rule")
+	return nil
+}
+
+func (service *VPCService) getorCreateAVIGroup(orgId string, projectId string, vpcId string, groupId string) (*model.Group, error) {
+	group, err := service.getAVIGroup(orgId, projectId, vpcId, groupId)
+	if err != nil {
+		log.Info("create avi group", "group", group)
+		group, err = service.createAVIGroup(orgId, projectId, vpcId, groupId)
+		if err != nil {
+			log.Error(err, "failed to create avi group", "group", groupId)
+		}
+	}
+	return group, err
+}
+
+func (service *VPCService) createAVIGroup(orgId string, projectId string, vpcId string, groupId string) (*model.Group, error) {
+	var group model.Group
+	tags := []model.Tag{
+		{
+			Scope: pointy.String(common.TagScopeCluster),
+			Tag:   pointy.String(service.NSXConfig.Cluster),
+		},
+		{
+			Scope: pointy.String(common.TagScopeVersion),
+			Tag:   pointy.String(strings.Join(common.TagValueVersion, ".")),
+		},
+		{
+			Scope: pointy.String(common.TagScopeVPCCRUID),
+			Tag:   pointy.String(vpcId),
+		},
+		{
+			Scope: pointy.String(common.TagScopeGroupType),
+			Tag:   pointy.String("avi"),
```

Comment:
please define a common var for grouptype like https://github.com/vmware-tanzu/nsx-operator/blob/vpc_dev/pkg/nsx/services/common/types.go#L63

Reply:
done